File: app/src/main/java/com/example/ali/easyrepo/TestActivity.java
package com.example.ali.easyrepo;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.easyrepolib.abstracts.GRepo;
import com.example.easyrepolib.abstracts.onSaveCompleted;
import com.example.easyrepolib.repos.BitmapDAO;
import com.example.easyrepolib.repos.ByteDAO;
import com.example.easyrepolib.repos.ObjectDAO;
import com.example.easyrepolib.repos.SafeBox;
import com.example.easyrepolib.repos.StringDAO;
import com.example.easyrepolib.security.DeviceKeyGenerator;
import com.example.easyrepolib.sqlite.KeyValDb;

import java.io.File;
import java.util.List;

public class TestActivity extends AppCompatActivity {

    private TextView log;
    private ImageView image;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_test);
        log = findViewById(R.id.loger);
        image = findViewById(R.id.image);

        //run tests
//        BitmapTest();
        StringTest();
        ByteTest();
        ObjectTest();
        GenericDAOTest();
    }

    private void BitmapTest() {

        final BitmapDAO bitmapDAO = new BitmapDAO(this, GRepo.Mode.LOCAL);

        Bitmap mBitmap = getBitmapForTest();

        bitmapDAO.SaveAsync("filename", mBitmap, this, new onSaveCompleted() {
            @Override
            public void onSaveComplete() {
                bitmapDAO.LoadAsync("filename", TestActivity.this, new BitmapDAO.OnBitmapLoad() {
                    @Override
                    public void onBitmapLoad(final Bitmap bitmap) {
                        image.setImageBitmap(bitmap);
                    }
                });
            }
        });
    }

    private void ByteTest() {

        ByteDAO byteDAO = new ByteDAO(this, GRepo.Mode.LOCAL);

        byte[] bytes = new byte[4];

        bytes[0] = 1;
        bytes[1] = 1;
        bytes[2] = 0;
        bytes[3] = 1;

        if (byteDAO.CheckExist("test")) {
            log("test is exist");
        } else {
            log("test is not exist ");
        }

        log("saving...");
        byteDAO.Save("test", bytes);
        log("saved");

        if (byteDAO.CheckExist("test")) {
            log("test is exist");
        } else {
            log("test is not exist ");
        }

        log("loading...");
        byte[] loadedBytes = byteDAO.Load("test");
        log("loaded");

        log("READED:" + loadedBytes);

        log("bytes");
        for (File file : byteDAO.GetAll()) {
            log("--" + file.getName());
        }


        log("removing...");
        byteDAO.Remove("test");
        log("removed");

        if (byteDAO.CheckExist("test")) {
            log("test is exist");
        } else {
            log("test is not exist ");
        }

        log("------------string test done-----------");

    }

    private void StringTest() {

        StringDAO stringDAO = new StringDAO(this, GRepo.Mode.LOCAL);

        String string = "write this file";

        if (stringDAO.CheckExist("test")) {
            log("test is exist");
        } else {
            log("test is not exist ");
        }

        log("saving...");
        stringDAO.Save("test", string);
        log("saved");

        if (stringDAO.CheckExist("test")) {
            log("test is exist");
        } else {
            log("test is not exist ");
        }

        log("loading...");
        String loadedString = stringDAO.Load("test");
        log("loaded");

        log("READED:" + loadedString);

        log("strings");
        for (File file : stringDAO.GetAll()) {
            log("--" + file.getName());
        }

        stringDAO.RemoveAll();

        log("strings");
        for (File file : stringDAO.GetAll()) {
            log("--" + file.getName());
        }

        log("removing...");
        stringDAO.Remove("test");
        log("removed");

        if (stringDAO.CheckExist("test")) {
            log("test is exist");
        } else {
            log("test is not exist ");
        }

        log("------------string test done-----------");

    }

    private void ObjectTest() {
        ObjectDAO o = new ObjectDAO(this, GRepo.Mode.LOCAL);

        User u = new User();
        u.name = "ali";

        o.Save("user",u);

        o.Load("user",User.class);

    }

    private void GenericDAOTest(){

        UserDao db = new UserDao(this);

        db.Drop();

        User u = new User();

        u.age = 18;
        u.name = "ali";
        u.lName = "ghahremani";

        db.Insert(u);
        u.name = "hassan";
        db.Insert(u);
        db.Insert(u);

        List<User> users = db.getAll();

        List<User> alis = db.getWithCondition(new KeyValDb.Condition() {
            @Override
            public boolean IsConditionTrue(Object object) {
                return  ((User) object).name.equals("ali");
            }
        });
    }

    private void TestSafeBox(){
        String key = DeviceKeyGenerator.Generate(this);
        SafeBox safeBox = new SafeBox(this,key);

        safeBox.Save("password","myPassword");

        safeBox.Load("password");
    }

    private void log(String msg) {
        log.append(msg + "\n");
    }

    private Bitmap getBitmapForTest() {
        return BitmapFactory.decodeResource(getResources(), R.mipmap.ic_launcher);
    }
}

File: app/src/main/java/com/example/ali/easyrepo/TestActivity.kt
package com.example.ali.easyrepo

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.util.Log
import android.widget.ImageView
import android.widget.TextView
import com.example.easyrepolib.repos.*
import com.example.easyrepolib.security.DeviceKeyGenerator.Generate

class TestActivity : Activity() {
    private var log: TextView? = null
    private var image: ImageView? = null

    // Tools
    private val bitmapForTest: Bitmap
        get() = BitmapFactory.decodeResource(resources, R.mipmap.ic_launcher)
    private fun log(msg: String){
        Log.d("test",msg)
        log!!.append("$msg\n")
    }
    private fun logTitle(title:String) = log("\n----- $title -----")
    private fun test(name:String,bool:Boolean) = log(name+"\t\t\t" + if (bool) " passed!" else " failed! ")
    private fun test(name:String,a:Any,b:Any) = test(name,a==b)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_test)
        log = findViewById(R.id.logger)
        image = findViewById(R.id.image)

//        bitmapTest()
        byteTest()
        stringTest()
        objectTest()
        genericDAOTest()
        testSafeBox()
    }

    private fun bitmapTest() {
        logTitle("bitmap test")

        val bitmapDAO = BitmapDAO(this, RootMode.LOCAL)

        // Sync
        bitmapDAO.save("testBitmap.bmp",bitmapForTest,100)
        test("testBitmap.bmp",bitmapDAO.checkExist("testBitmap.bmp"))

        bitmapDAO.save("testBitmap50.bmp",bitmapForTest,50)
        test("testBitmap50.bmp",bitmapDAO.checkExist("testBitmap50.bmp"))

        bitmapDAO.save("testBitmap10px.bmp",bitmapForTest, width=10, height=10)
        test("testBitmap10px.bmp",bitmapDAO.checkExist("testBitmap10px.bmp"))

        // Async
        bitmapDAO.saveAsync("testBitmapAsync.bmp", bitmapForTest, callback = {
            test("testBitmapAsync.bmp",bitmapDAO.checkExist("testBitmapAsync.bmp"))
        })
        Thread.sleep(1000)
        log("bitmap tests all done!")
    }

    private fun byteTest() {

        logTitle("byte test")

        val byteDAO = ByteDAO(this, RootMode.LOCAL)
        val bytes = ByteArray(4)
        bytes[0] = 1
        bytes[1] = 1
        bytes[2] = 0
        bytes[3] = 1

        // saving
        byteDAO.save("test", bytes)
        test("saving",byteDAO.checkExist("test"))

        // loading
        val loadedBytes: ByteArray = byteDAO.load("test")!!
        test("byte eq:",loadedBytes[0],bytes[0])

        // removing
        byteDAO.remove("test")
        test("removing",!byteDAO.checkExist("test"))
    }

    private fun stringTest() {

        logTitle("string test")

        val stringDAO = StringDAO(this, RootMode.LOCAL)
        val string = "write this file"

        //saving
        stringDAO.save("test", string)
        test("saving",stringDAO.checkExist("test"))

        // loading
        val loadedString: String = stringDAO.load("test")
        test("byte eq:",loadedString,string)

        // removing
        stringDAO.remove("test")
        test("removing",!stringDAO.checkExist("test"))

        // default
        test("default",stringDAO.load("test","def"),"def")
    }

    private fun objectTest() {

        logTitle("object test")

        val dao = ObjectDAO(this, RootMode.LOCAL)
        val user = User("id","pass","ali",18,"admin",1000)

        dao.save("test", user)
        test("saving",dao.checkExist("test"))

        // loading
        val loaded: User = dao.load("test",User::class.java) as User
        test("equality:",loaded.name,user.name)

        // removing
        dao.remove("test")
        test("removing",!dao.checkExist("test"))
    }

    private fun genericDAOTest() {

        logTitle("generic test")

        val users = User.getRepo(this)

        val u = User("id","pass","ali",18,"admin",1000)

        //insert
        users.insert(u)

        test("insert",users.getById("id").name,"ali")

        //read
        test("read",users.all.size,1)

        //update
        u.name = "hasan"
        users.update(u)
        test("update",users.getById(u.id).name,"hasan")

        //delete
        users.delete(u.id)
        test("delete",users.all.size,0)

        users.drop()
    }

    private fun testSafeBox() {
        logTitle("safebox test")
        val key = Generate(this)
        val safeBox = SafeBox(this, key)

        safeBox.save("password", "myPassword")
        var readed = safeBox.load("password")
        test("read eq",readed=="myPassword")

        safeBox.save("password", "سلام")
        readed = safeBox.load("password")
        test("utf8",readed=="سلام")
    }
}